swarm/network/stream/snapshot_retrieval_test.go

```diff
@@ -0,0 +1,378 @@
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+package stream
+
+import (
+	"context"
+	//	crand "crypto/rand"
+	"fmt"
+	//"io"
+	"math/rand"
+	//"sync"
+	"testing"
+	"time"
+
+	"github.com/ethereum/go-ethereum/log"
+	//"github.com/ethereum/go-ethereum/p2p"
+	"github.com/ethereum/go-ethereum/p2p/discover"
+	"github.com/ethereum/go-ethereum/p2p/simulations"
+	//"github.com/ethereum/go-ethereum/p2p/simulations/adapters"
+	//"github.com/ethereum/go-ethereum/pot"
+	//"github.com/ethereum/go-ethereum/rpc"
+	"github.com/ethereum/go-ethereum/swarm/network"
+	streamTesting "github.com/ethereum/go-ethereum/swarm/network/stream/testing"
+	"github.com/ethereum/go-ethereum/swarm/storage"
+)
+
+func initRetrievalTest() {
+	toAddr = func(id discover.NodeID) *network.BzzAddr {
+		addr := network.NewAddrFromNodeID(id)
+		return addr
+	}
+	createStoreFunc = createTestLocalStorageForId
+	//local stores
+	stores = make(map[discover.NodeID]storage.ChunkStore)
+	//data directories for each node and store
+	datadirs = make(map[discover.NodeID]string)
+	//deliveries for each node
+	deliveries = make(map[discover.NodeID]*Delivery)
+	getRetrieveFunc = func(id discover.NodeID) func(chunk *storage.Chunk) error {
+		return func(chunk *storage.Chunk) error {
+			skipCheck := true
+			//fmt.Println(fmt.Sprintf("-- %s", id))
+			return deliveries[id].RequestFromPeers(chunk.Key[:], skipCheck)
+		}
+	}
+	//registries, map of discover.NodeID to its streamer
+	registries = make(map[discover.NodeID]*TestRegistry)
+	//channel to wait for peers connected
+	//not needed for this test but required from common_test for NewStreamService
+	waitPeerErrC = make(chan error)
+	//also not needed for this test but required for NewStreamService
+	peerCount = func(id discover.NodeID) int {
+		if ids[0] == id || ids[len(ids)-1] == id {
+			return 1
+		}
+		return 2
+	}
+}
+
+func TestRetrieval(t *testing.T) {
+
+	if *nodes != 0 && *chunks != 0 {
+		retrievalTest(t, *chunks, *nodes)
+	} else {
+		var nodeCnt []int
+		var chnkCnt []int
+		if *longrunning {
+			nodeCnt = []int{16, 32, 128}
+			chnkCnt = []int{4, 32, 256}
+		} else {
+			nodeCnt = []int{16}
+			chnkCnt = []int{32}
+		}
+		for _, n := range nodeCnt {
+			for _, c := range chnkCnt {
+				retrievalTest(t, c, n)
+			}
+		}
+	}
+}
+
+func retrievalTest(t *testing.T, chunkCount int, nodeCount int) {
+	//test live and NO history
+	log.Info("Testing live and no history", "chunkCount", chunkCount, "nodeCount", nodeCount)
+	live = true
```

Comment:
why are these global and not params to `runRetrievalTest`?

Reply:
They need to be global due to the (current) `StartSyncing` test implementation, as they are needed there for setting up the stream, thus they were globally set in `snapshot_sync_test`. They could be additional RPC call params though.